src/main/scala/tech/cryptonomic/conseil/generic/chain/MultiFetchDecoding.scala

```diff
@@ -0,0 +1,136 @@
+package tech.cryptonomic.conseil.generic.chain
+
+import cats._
+import cats.data.Kleisli
+import cats.syntax.semigroupal._
+
+/** Contract for a generic combination of operations that get encoded data (usually json) from
+  * some resource (e.g. the tezos node), then converts that to a value, batching multiple requests together.
+  * The returned values are tupled with the corresponding input
+  * The interface assumes some kind of higher-order typed effect `F[_]` is part
+  * of the operation execution, and represents everything parametrically, to leave
+  * room for future evolutions of the code protocol.
+  * Thus the required implementation is in the form of a `cats.data.Kleisli` type,
+  * parametric on effect `F`, input `In`, output `Out`. As needed, the in/out types are
+  * wrapped in a `T[_]` type constructor that represents multiple values (e.g. a collection type).
+  *
+  * @tparam F the higher-order effect type (e.g. `Try`, `Future`, `IO`, `Either`)
+  * @tparam T a traversable type (used to wrap mutliple requests and corresponding return values)
+  */
+trait MultiFetchDecoding[F[_], T[_]] {
+  /** the input type, e.g. ids of data */
+  type In
+  /** the output type, e.g. the decoded block data */
+  type Out
+  /** the encoded representation type used e.g. some Json representation */
+  type Encoded
+
+  /** an effectful function from a collection of inputs `T[In]`
+    * to the collection of encoded values, tupled with the corresponding input `T[(In, Encoded)]`
+    */
+  def fetchBatch: Kleisli[F, T[In], T[(In, Encoded)]]
+
+  /** an effectful function that decodes the json value to an output `Out` */
+  def decodeData: Kleisli[F, Encoded, Out]
+
+  /** using the combination of the provided functions we fetch all data with this general outline:
+    * 1. get json values from the node, given a traversable input collection
+    * 2. keep the correlation with the input and convert every json to a corresponding output
+    * 3. return the traversable collection containing paired input and output values, all wrapped in the effect F
+    */
+  def fetch(implicit app: Monad[F], tr: Traverse[T]): Kleisli[F, T[In], T[(In, Out)]] =
+    fetchBatch.andThen(json => decodeData.second[In].traverse(json))
```

Comment:
For my education..since so much set up was done with Kleislis to potentially set up a neat composition, how come there is no 'higher-order Kleisli' that is able to operate on a traversible of effects so this line simply becomes something like:

`fetchBatch.andThenTraverse(decodeData..)`

?

Reply:
I would say that they are orthogonal concerns, composition of functions (andThen) wrt traversing a structure.
Since the result of the kleisli application (i.e. the result of a function with effect) might have a very general form, you could apply many possible computations to that.
e.g. it could be optional, monoidal, monadic, applicative, alternative, ... each of those type classes provides many operations, on the "content" of the result, and providing custom composition for the kleisli own type classes (arrow, alternative, functor, contravariant-functor, ...) with those of the content would lead to an explosion of custom operators.
This would also be limiting, as new type classes will be defined, and no ad-hoc combination would be available for those.

If you look at `Future.sequence` and `Future.traverse`, that is an ad-hoc commodity feature, that doesn't scale much, because after a while you start asking yourself why only those and why only on Future and not, say Option...
This is my own explanation, the fact that you want to keep concerns separated without tying up separate type classes.

As a side-note, a different reasoning applies if having an additional constraint on the, say, result of a kleisli operation, might allow you to create powerful shortcuts, like having a monoid/semigroup (that is, you can combine many results of the same type to a single value).
This is used if you got a collection of monoid things, that you can immediately reduce to a single value.
Possibly the andThenTraverse combinator is not so common to prove useful yet.